### DISCUSSION TITLE
Do we need an API to define that the onChange validator is launched also on mount?

### DISCUSSION THREAD

**Discussion by Balastrong** [TanStack/form#1005] [Ideas]
In some cases we might want our form to run the onChange validator as soon as the component mounts, do we need a specific API to handle that?

After #726 (version 0.34.4) you can easily achieve that by defining your validating function and passing it to both validators

```ts
function myValidator () { ... }

validators={{
    onMount: myValidator,
    onChange: myValidator,
}}
```

This will set `canSubmit` to false if there are errors but will not set `isTouched` to fields so the user can control if they want to show the error or not.

(See [Stackblitz with zod](https://stackblitz.com/edit/tanstack-form-rjkvj7?file=src%2Findex.tsx))

Is that enough or we need something at Form/Field level to declare that? Suggestions and ideas are welcome!

**Comment by rik-iso** [comment]
I think so. I'm having to do this right now (little cludgy workaround using radashi's [objectify](https://radashi.js.org/reference/array/objectify/), but obviously another helper could be written):

```tsx
validators={objectify(['onMount', 'onBlur'], identity, () =>
   v.pipe(v.number(), v.minValue(1), v.maxValue(999))
)}
```

I don't think it's particularly ideal.

**Reply by Balastrong** [reply]
Using an extra library just for that definitely seems overkill and shouldn't be needed.

Does it work with something like that?
```ts
const numberSchema = v.pipe(v.number(), v.minValue(1), v.maxValue(999))

validators={{
  onMount: numberSchema,
  onChange: numberSchema,
}}
```

**Reply by rik-iso** [reply]
Oh yeah it does, I was just using that library because we already had it.

**Comment by luap2703** [comment]
We definitely need an api for this! :)

We are running quite complex forms with plenty of fields (thus everything is quite typescript heavy). It would be perhaps best to have a flag that is onAny: () => {} and/or directly passing a function into validators={} that does that job.

**Reply by AhsanRaza007** [reply]
any updates on this?

**Comment by tjbrennan** [comment]
I also want to point out that using two validators this way results in duplicate error messages for empty required fields where the user has clicked into an input and clicked away without entering a value.

**Reply by pleunv** [reply]
Just ran into this as well. Not sure if this is a regression of #726 but definitely unexpected. Have to check the actual values for now to determine validity on initial render.
